### PR TITLE
Cleanup some method tests in System.Dynamic.Runtime

### PR DESCRIPTION
**Pull request by hughbe**
- Consolidate common test data
- Rename tests to make sense
- Extract common code
- Deobfuscate some tests
Contributes to #3133, #915, #516

/cc @VSadov @stephentoub

### CODE REVIEW

File: src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs [hunk 1]
```diff
@@ -239,1557 +189,484 @@ public U Method_ReturnsU()
...
+        [Fact]
+        public static void ImplicitlyTypedVariableInitializer()
         {
-            dynamic mc = new MemberClass<int>();
-            int ret = mc.Method_ReturnsT<string>(string.Empty, "ab", null);
-            dynamic d1 = string.Empty;
-            dynamic d2 = "ab";
-            ret += mc.Method_ReturnsT(d1, d2, null);
-            return ret;
+            dynamic d = new GenericClass<string>();
```
Comment (VSadov) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: Is this supposed to be equivalent of ManagedTests.DynamicCSharp.Conformance.dynamic.context.method.genmethod.genclass.genclass003.genclass003.MainMethod ?

File: src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs [hunk 2]
```diff
@@ -239,1557 +189,484 @@ public U Method_ReturnsU()
...
-            Assert.Equal(0, MainMethod());
+            dynamic dy1 = new GenericClass<int>();
+            long result1 = (int)dy1.Method_ReturnsDynamic(1); // implicit
+            dynamic dy2 = new GenericClass<long>();
+            int result2 = (int)dy2.Method_ReturnsDynamic(1L); // explicit
+            Assert.Equal(result1, result2);
         }
 
-        public static int MainMethod()
+        [Fact]
```
Comment (VSadov) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: Here we had a test with ```    // <Title> Tests generic class generic method used in implicit operator.</Title>```
Is it possible to keep tests in the same order and keep some identifying information like "Title" comments. 

It is very hard to follow these changes without any way of referencing old test scenarios and the new ones.

Comment (hughbe) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: I'm sorry - I probably should have tried to make this easier to review - I did this cleanup a while back. Looking back on it, the diff's pretty horrible.
If it helps, the ordering of tests hasn't changed at all.
In many cases, the titles are incorrect - I have tried to include information from titles where they are correct in the title of the test.

For instance, this test is called `ImplicitExplicitOperator()`

Comment (hughbe) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: You're probably an infinitely better code reviewer than I am, but I have a suggestion.
Bearing in bind that I've left the tests ordered and they have a similar name you could try a side by side comparison:
e.g. split the screen between:
- https://github.com/hughbe/corefx/blob/fb169beb35da1e022d2c42163012c9b905ea2ab5/src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs
- https://github.com/dotnet/corefx/blob/master/src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs

This would involve a bunch of scrolling, but is better IMO than copy and pasting trying to find each test individual, right?

File: src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs (hunk 1 above)
Comment (hughbe) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: Yes. I assume I made a git error or something, as the method body after should be:
```
dynamic s_mc = new GenericClass<string>()
Assert.Equal(0, (int)s_mc.Method_ReturnsU<int>(null));
```

File: src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs [hunk 3]
```diff
@@ -239,1557 +189,484 @@ public U Method_ReturnsU()
...
-            Assert.Equal(0, MainMethod());
+            dynamic mc = new GenericClass<string>();
+            Assert.Null((string)mc.Method_ReturnsT<long>(1.11f, string.Empty, 1L));
         }
 
-        public static int MainMethod()
+        [Fact]
+        public static void StaticMethodBody3()
         {
-            var loc = (int)s_mc.Method_ReturnsU<int>(null);
```
Comment (VSadov) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: >var loc = (int)s_mc.Method_ReturnsU<int>(null); [](start = 12, length = 47)

where did this go?   (```var loc = (int)s_mc.Method_ReturnsU<int>(null);```)

File: src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs [hunk 4]
```diff
@@ -239,1557 +189,484 @@ public U Method_ReturnsU()
...
+        [Fact]
+        public static void AnonymousMethod()
+        {
+            dynamic d = new GenericClass<string>.NestedGenericClass<int>();
+            Func<string, int> func = delegate (string arg)
             {
-                dynamic dy = new MemberClass<InnerTest1>();
-                return new InnerTest2()
-                {
-                    field = (int)dy.Method_ReturnsU<int>(t1.field + 1)
```
Comment (VSadov) [inline src/System.Dynamic.Runtime/tests/Dynamic.Context/Conformance.dynamic.context.method.genmethod.genclass.cs]: >field = (int)dy.Method_ReturnsU<int>(t1.field + 1) [](start = 20, length = 50)

Can't find equivalent to this -  ```field = (int)dy.Method_ReturnsU<int>(t1.field + 1)```

**Comment by hughbe** [comment]
I think you're right. Sorry for creating work for you - I probably wouldn't have submitted the PR in this now.

What I'm gonna do is split this up into multiple PRs (1 for each file).

I'll take some time to add the original comment name. But I would like to push back on your first point. I think its more descriptive to make the test title itself.

E.g. `

**Comment by stephentoub** [comment]
Thanks, @hughbe.  Given your comments, I'm closing this, and you can open replacements.
